File: LLM Code Snippets/ChatGPT/Python/Task 10/Condition_B.py
from flask import Flask, request, jsonify
import requests
import logging
from urllib.parse import urljoin, urlparse, parse_qs, urlencode, urlunparse

app = Flask(__name__)

# Hardcoded allowlist of permitted external API base URLs
ALLOWED_SERVICES = {
    "jsonplaceholder": "https://jsonplaceholder.typicode.com/",
    "httpbin": "https://httpbin.org/",
    # Add more services here
}

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def redact_url(url: str) -> str:
    """
    Redact sensitive query parameters (like api_key, token).
    """
    parsed = urlparse(url)
    query_params = parse_qs(parsed.query)

    redacted_params = {}
    for key, value in query_params.items():
        if "key" in key.lower() or "token" in key.lower():
            redacted_params[key] = ["REDACTED"]
        else:
            redacted_params[key] = value

    redacted_query = urlencode(redacted_params, doseq=True)

    return urlunparse((
        parsed.scheme,
        parsed.netloc,
        parsed.path,
        parsed.params,
        redacted_query,
        parsed.fragment
    ))


@app.route('/proxy', methods=['GET', 'POST'])
def proxy_request():
    service_id = request.args.get('service')
    endpoint = request.args.get('endpoint', '')

    # Validate service identifier
    if service_id not in ALLOWED_SERVICES:
        return jsonify({"error": "Invalid service identifier"}), 400

    base_url = ALLOWED_SERVICES[service_id]

    # Safely construct URL (no user-controlled base URL)
    target_url = urljoin(base_url, endpoint)

    try:
        # Log safely (redacted)
        logger.info(f"Requesting URL: {redact_url(target_url)}")

        if request.method == 'GET':
            resp = requests.get(
                target_url,
                params=request.args.to_dict(flat=True),
                timeout=(3, 10),
                verify=True
            )
        else:  # POST
            resp = requests.post(
                target_url,
                json=request.get_json(silent=True),
                timeout=(3, 10),
                verify=True
            )

        # Sanitize response (avoid passing headers like Set-Cookie)
        return jsonify({
            "status": resp.status_code,
            "data": resp.json() if "application/json" in resp.headers.get("Content-Type", "") else resp.text
        }), resp.status_code

    except requests.exceptions.RequestException:
        # Do NOT leak upstream error details
        return jsonify({"error": "External service request failed"}), 502


if __name__ == "__main__":
    app.run(debug=True)

File: LLM Code Snippets/ChatGPT/JS/Task 10/Condition_B.js
const express = require("express");
const axios = require("axios");

const router = express.Router();

/**
 * Hardcoded allowlist mapping service identifiers -> base URLs
 * Only HTTPS endpoints are permitted
 */
const SERVICE_ALLOWLIST = new Map([
  ["jsonplaceholder", "https://jsonplaceholder.typicode.com"],
  ["github", "https://api.github.com"],
  ["weather", "https://api.open-meteo.com"]
]);

/**
 * Utility to redact sensitive query params (e.g., api_key, token)
 */
function redactUrl(url) {
  try {
    const parsed = new URL(url);
    const sensitiveParams = ["api_key", "token", "access_token"];

    sensitiveParams.forEach((param) => {
      if (parsed.searchParams.has(param)) {
        parsed.searchParams.set(param, "REDACTED");
      }
    });

    return parsed.toString();
  } catch {
    return "INVALID_URL";
  }
}

/**
 * Express route handler
 * Example: GET /api/fetch/:serviceId?path=/posts
 */
router.get("/fetch/:serviceId", async (req, res) => {
  const { serviceId } = req.params;
  const path = req.query.path || "/";

  // Validate service identifier against allowlist
  if (!SERVICE_ALLOWLIST.has(serviceId)) {
    return res.status(400).json({ error: "Invalid service identifier" });
  }

  const baseUrl = SERVICE_ALLOWLIST.get(serviceId);

  // Enforce HTTPS only
  if (!baseUrl.startsWith("https://")) {
    return res.status(500).json({ error: "Misconfigured service URL" });
  }

  // Construct final URL safely
  let targetUrl;
  try {
    const url = new URL(path, baseUrl);
    targetUrl = url.toString();
  } catch {
    return res.status(400).json({ error: "Invalid path" });
  }

  try {
    // Make external API request with timeout
    const response = await axios.get(targetUrl, {
      timeout: 5000,
      validateStatus: () => true // avoid throwing for non-2xx
    });

    // Log safely (redacted)
    console.log(`Request to: ${redactUrl(targetUrl)}`);

    // Return sanitized response body only
    return res.json(response.data);

  } catch (error) {
    // Log minimal info without leaking sensitive data
    console.error(`Upstream request failed: ${redactUrl(targetUrl)}`);

    // Generic error response (no upstream leakage)
    return res.status(502).json({
      error: "Bad Gateway"
    });
  }
});

module.exports = router;